x-pack/plugins/enterprise_search/public/applications/app_search/components/credentials/credentials_logic.test.ts

```diff
@@ -0,0 +1,1006 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { resetContext } from 'kea';
+
+import { CredentialsLogic } from './credentials_logic';
+import { ADMIN, PRIVATE } from '../../constants/credentials';
+
+jest.mock('../../../shared/http', () => ({
+  HttpLogic: { values: { http: { get: jest.fn(), delete: jest.fn() } } },
+}));
+import { HttpLogic } from '../../../shared/http';
+
+function mount(defaults?) {
+  if (!defaults) {
+    resetContext({});
+  } else {
+    resetContext({
+      // the enterprise_search.app_search.credentials namespace is set in `path` in credentials_logic
+      defaults: {
+        enterprise_search: {
+          app_search: {
+            credentials: {
+              ...defaults,
+            },
+          },
+        },
+      },
+    });
+  }
+  CredentialsLogic.mount();
+}
+
+const newToken = {
+  id: '1',
+  name: 'myToken',
+  type: PRIVATE,
+  read: true,
+  write: true,
+  access_all_engines: true,
+};
+
+describe('CredentialsLogic', () => {
+  const DEFAULT_VALUES = {
+    activeApiToken: {
+      name: '',
+      type: PRIVATE,
+      read: true,
+      write: true,
+      access_all_engines: true,
+    },
+    activeApiTokenIsExisting: false,
+    activeApiTokenRawName: '',
+    apiTokens: [],
+    apiUrl: '',
+    dataLoading: true,
+    engines: [],
+    formErrors: [],
+    isCredentialsDataComplete: false,
+    isCredentialsDetailsComplete: false,
+    meta: '',
+    nameInputBlurred: false,
+    showCredentialsForm: false,
+  };
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('has expected default values', () => {
+    mount();
+    expect(CredentialsLogic.values).toEqual(DEFAULT_VALUES);
+  });
+
+  describe('addEngineName', () => {
+    describe('activeApiToken', () => {
+      it("should add an engine to the active api token's engine list", () => {
+        mount({
+          activeApiToken: {
+            ...newToken,
+            engines: ['someEngine'],
+          },
+        });
+        CredentialsLogic.actions.addEngineName('newEngine');
+        expect(CredentialsLogic.values.activeApiToken).toEqual({
+          ...newToken,
+          engines: ['someEngine', 'newEngine'],
+        });
+      });
+
+      it("create a new engines list if one doesn't exist", () => {
+        mount({
+          activeApiToken: {
+            ...newToken,
+            engines: undefined,
+          },
+        });
+        CredentialsLogic.actions.addEngineName('newEngine');
+        expect(CredentialsLogic.values.activeApiToken).toEqual({
+          ...newToken,
+          engines: ['newEngine'],
+        });
+      });
+    });
+
+    it('should not change any other values', () => {
```

Comment:
I added one of these tests to each action. My reasoning is this:

1. As I was testing, it helped ensure for me that I had covered all of the necessary values with tests. It will also serve as a reminder to add test coverage as we make updates as well.
2. Avoid unintentional values changes in actions

Reply:
Follow up to my comment here: https://github.com/elastic/kibana/pull/77626#discussion_r490634386

I strongly think we should not have tests specifically for `should not change any other values`. It's hard to read/understand, doesn't add anything significantly to line coverage, and bloats the file. That said, it's fully possible to check for this in your active tests that check for updates. Simply do this:

```js
// example assertion for "should remove an engine from the active api token's engine list" test

        expect(CredentialsLogic.values).toEqual({
          ...DEFAULT_VALUES,
          activeApiToken: {
            ...newToken,
            engines: ['anotherEngine'],
          },
        });
```

Instead of:

https://github.com/elastic/kibana/blob/91392aec8d752972cd94ae930f593da43fd2d5e7/x-pack/plugins/enterprise_search/public/applications/app_search/components/credentials/credentials_logic.test.ts#L144-L147

... and bam, you've proved other values haven't changed at the same time you have asserted your expected values do change, in a fixed 3 lines instead of 9+ 🤷 (also cuts down on test times / time spent mounting extra tests, etc.).

FWIW, we also don't have to check the full` expect(SomeLogic.values)` for _every_ action>value block, just the first one - follow up edge case/conditional branch tests can simply check `SomeLogic.values.someValue` after that since the first test asserts unchanged values for us.

Reply:
It gets a little weird when you have a bunch of values that change when you fire an action though, because then you can't just use DEFAULT_VALUES.

```
        expect(CredentialsLogic.values).toEqual({
          ...DEFAULT_VALUES,
          someOtherThing: 'foo',
          anotherValueThatChanged: 'bar',
          activeApiToken: {
            ...newToken,
            engines: ['anotherEngine'],
          },
        });
```

You'd have to put in assertions for `anotherValueThatChanged` and `someOtherThing` as well, as seen above. At which point it's like, well I have a separate test for `anotherValueThatChanged` and `someOtherThing` below, why do I need both if it's being asserted here?

One way to avoid that is to use `expect.any` to make it clear that that is not what's under test currently, and we don't care what is provided here.

```
        expect(CredentialsLogic.values).toEqual({
          ...DEFAULT_VALUES,
          someOtherThing: expect.any(string),
          anotherValueThatChanged: expect.any(string),
          activeApiToken: {
            ...newToken,
            engines: ['anotherEngine'],
          },
        });
```

Which is what I arrived at, and I ultimately thought it would be clearer to just move that into a separate test so it's explicit what is being tested. I may have taken it a step too far though.

I don't have a strong opinion here. I think it IS important to assert that the other values do not change somehow though. The completed updated state (the values) are returned from every action and hence part of the public interface of that action and should be asserted. If you prefer the inline assertion on the first test, I find that acceptable.

I wonder if something like this would work....

```
const values = {
   ...DEFAULT_VALUES,
   someOtherThing: expect.any(string),
   anotherValueThatChanged: expect.any(string),
   activeApiToken: expect.any(object)
}

....          
        expect(CredentialsLogic.values).toEqual({
          ...values,
          activeApiToken: {
            ...newToken,
            engines: ['anotherEngine'],
          },
        });

...
        expect(CredentialsLogic.values).toEqual({
          ...values,
          anotherValueThatChanged: 'foo'
        });

...

        expect(CredentialsLogic.values).toEqual({
          ...values,
          someOtherThing: 'foo'
        });
```

Thoughts?

Reply:
You totally beat me to this lol, I spent a few more hrs last night reading through the test file and definitely get now why my proposed solution doesn't work well for all the test suites (esp. the ones dealing w/ a bunch of deeply nested obj children).

> You'd have to put in assertions for anotherValueThatChanged and someOtherThing as well, as seen above. At which point it's like, well I have a separate test for anotherValueThatChanged and someOtherThing below, why do I need both if it's being asserted here?

Definitely a good point, and also potentially a indication that (for _some_, definitely not all) tests, it may be worth considering whether it would be easier to just group all the changes values in a combined single test/assertion, rather than splitting them out 🤔 That being said, I totally get why not though for a lot of the more complex tests - also this is not me requesting a change to the current PR, just potentially food for thought for future logic tests.

All that being said, I also like the proposed compromise a lot! Would love to give it a shot - feel free to timebox though, if it starts taking too long to refactor or creates unexpected issues I'd also be fine leaving the current `should not change` tests as-is.

Reply:
6347c61d51a8c97b38bf9d0451f3d88ee22f4de0